### DISCUSSION TITLE
Track and show Mouse pointer coordinates within the panel

### DISCUSSION THREAD

**Discussion by motihat** [gapitio/gapit-htmlgraphics-panel#102] [Q&A] [answered]
Is there a way to track and display mouse coordinates within the htmlgraphics-panel.  If not possible to track continuously then just showing the X,Y coordinates on mouse click anywhere within the panel would suffice too. An simple example would be greatly appreciated. 
I am new user and a beginner in Javascript and HTML.

**Comment by ZuperZee** [comment, marked as answer]
You can "track" the mouse by using the [mousemove event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousemove_event) on any element.

```js
document.body.addEventListener("mousemove", (event) => {console.log(event.x, event.y)}); // The entire page
htmlNode.lastElementChild.addEventListener("mousemove", (event) => {console.log(event.x, event.y)}); // The panel
```

Because of some "bad timings" with mousemove, I found it more consistent to add a timeout (with 0 delay) in the mousemove event handler.
```js
htmlNode.lastElementChild.addEventListener("mousemove", (event) => {setTimeout(() => console.log(event.x, event.y), 0)});
```

Here is an example:
_You can import it directly in the panel https://gapit-htmlgraphics-panel.gapit.io/docs/guides/how-to-import-export/_

`Panel options`

```json
{
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": "* {\n  font-family: Open Sans;\n}\n\n#mouse-tracker {\n  background-color: #3D71D9;\n\n  width: 100px;\n  height: 100px;\n\n  position: absolute;\n\n  top: 50px;\n  left: 50px;\n\n  transform: translate(-50%, -20%);\n\n  display: flex;\n  justify-content: center;\n  align-items: center;\n}",
    "html": "<div>\n  <div id=\"mouse-tracker\">Hello</div>\n</div>",
    "renderOnMount": true,
    "onRender": "",
    "dynamicHtmlGraphics": false,
    "dynamicData": false,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "panelupdateOnMount": true,
    "onInitOnResize": false,
    "onInit": "const mouseTrackerElt = htmlNode.querySelector(\"#mouse-tracker\");\nconst mouseTrackerAreaElt = htmlNode.lastElementChild;\n\nmouseTrackerAreaElt.addEventListener(\"mousemove\", (e) => {\n  setTimeout(() => {\n    const {offsetX: x, offsetY: y} = e;\n\n    mouseTrackerElt.textContent = `X: ${x} Y: ${y}`;\n    mouseTrackerElt.style.left = x + \"px\";\n    mouseTrackerElt.style.top = y + \"px\";\n  }, 0)\n});\n"
}
```

Here is just the javascript code:

```js
const mouseTrackerElt = htmlNode.querySelector("#mouse-tracker");
const mouseTrackerAreaElt = htmlNode.lastElementChild;

mouseTrackerAreaElt.addEventListener("mousemove", (e) => {
  setTimeout(() => {
    const {offsetX: x, offsetY: y} = e;

    mouseTrackerElt.textContent = `X: ${x} Y: ${y}`;
    mouseTrackerElt.style.left = x + "px";
    mouseTrackerElt.style.top = y + "px";
  }, 0)
});
```

And some beginner tips:
https://developer.mozilla.org/en-US/ has a lot of good information and examples about javascript like https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener.
The `=>` is called an [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)
The [mousemove event](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) has a few different x and y properties, but the most useful ones are [offsetX](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/offsetX) and [pageX](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/pageX).

**Reply by motihat** [reply]
Thanks a lot @ZuperZee for the detailed example and pointers to additional resources.  That was exactly what I was looking for.
